nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py

```diff
@@ -0,0 +1,312 @@
+
+        req = SetEntityState.Request()
+        req.state.name = entity_name
+        req.state.pose.position.x = entity_pose.position.x
+        req.state.pose.position.y = entity_pose.position.y
+        req.state.pose.position.z = 0.0
+        req.state.pose.orientation.x = entity_pose.orientation.x
+        req.state.pose.orientation.y = entity_pose.orientation.y
+        req.state.pose.orientation.z = entity_pose.orientation.z
+        req.state.pose.orientation.w = entity_pose.orientation.w
+        future = self.set_entity_state.call_async(req)
+        while not future.done() and rclpy.ok():
+            sleep(0.1)
+        sleep(0.5)
+
+    def get_robot_pose(self):
+        while not self.get_entity_state.wait_for_service(timeout_sec=1.0):
+            print('get entity state service is not available...')
+        req = GetEntityState.Request()
+        req.name = 'turtlebot3_waffle'
+        future = self.get_entity_state.call_async(req)
+
+        while not future.done() and rclpy.ok():
+            sleep(0.01 / self.time_factor)
+
+        self.current_pose.position.x = future.result().state.pose.position.x
+        self.current_pose.position.y = future.result().state.pose.position.y
+        self.current_pose.position.z = future.result().state.pose.position.z
+        self.current_pose.orientation.x = future.result().state.pose.orientation.x
+        self.current_pose.orientation.y = future.result().state.pose.orientation.y
+        self.current_pose.orientation.z = future.result().state.pose.orientation.z
+        self.current_pose.orientation.w = future.result().state.pose.orientation.w
+
+    def get_random_pose(self):
+        random_pose = Pose()
+        yaw = random.uniform(0, pi * 2)
+
+        random_pose.position.x = random.uniform(-3, 3)
+        random_pose.position.y = random.uniform(-3, 3)
+        random_pose.position.z = 0.0
+        random_pose.orientation.x = 0.0
+        random_pose.orientation.y = 0.0
+        random_pose.orientation.z = sin(yaw * 0.5)
+        random_pose.orientation.w = cos(yaw * 0.5)
+        return random_pose
+
+    def sq_distance_to_goal(self):
+        self.get_robot_pose()
+        dx = self.goal_pose.position.x - self.current_pose.position.x
+        dy = self.goal_pose.position.y - self.current_pose.position.y
+        return dx * dx + dy * dy
+
+    def get_heading(self):
+        goal_angle = math.atan2(self.goal_pose.position.y - self.current_pose.position.y,
+                                self.goal_pose.position.x - self.current_pose.position.x)
+
+        current_yaw = self.get_yaw(self.current_pose)
+        heading = goal_angle - current_yaw
+
+        if heading > pi:
+            heading -= 2 * pi
+
+        elif heading < -pi:
+            heading += 2 * pi
+
+        return heading
+
+    def get_yaw(self, q):
+        yaw = atan2(2.0 * (q.orientation.x * q.orientation.y + q.orientation.w * q.orientation.z),
```

Comment:
Is there really not a quaternion to yaw method in python to use?

Reply:
When I implemented this last month, the quaternion to yaw was not in python yet.  I need to check back to see if it is available now.

Reply:
Per this PR the tf2_ros is still not built for Python. https://github.com/ros2/geometry2/pull/99

Reply:
Sounds good